### PR TITLE
Improved hover detection for scatter plot fill `tonext*`

### PR DESCRIPTION
**Pull request by lumip**
Hover detection (and hover label placement) does not work correctly on scatter plot fills when using `tonextx`, `tonexty`, as demonstrated in #2165 . This is due to faulty construction of the hover detection polygons in these cases.

This PR contributes the following commits to address the issue:
- add tests for hover detection for `self`, `tonext`, `tonexty` and `tozery` based on test mocks `scatter_fill_corner_cases` and `scatter_fill_self_next`, paying special attention to corner cases where current hover detection fails to detect the correct trace ( 46287413889cbfed6c33ec48c9c8e3c01eadced2 )
- fix the faulty polygon construction that caused the misdetections reported in #2165 in `src/traces/scatter/plot.js`. However, this still results in some misdetections, especially for traces with curved edges or "jagged" shapes (e.g. `shape='vh'`), where the detection polygon construction will not follow the visual edge. ( be658dd7e52c5d5fb959f6c373404512737b38b9 )
- reimplement hover detection using the `isPointInFill` method of the `SVGElement` used to render the fill. This result in always correct detection of the hovered over trace. However, the `SVGElement` does not provide an easy means to compute hover label placement, so the detection polygons are still used for that. Close to curved edges, where polygons are not accurate, this can cause labels to end up in weird positions. ( 3eb473895518880bd630ba09c8318d6d27c92645 )
- change the label fallback placement (when polygon-based placement fails, see above) to be at the cursor instead, which to me seems more sensible and simplified the code. ( 63d181a29f0d0c05402dce413e02e53ecf137455 )

This is currently still not a perfect solution but I think this has reached a reasonable state now. I'm not sure whether the change in the last commit is desired but it seemed the more intuitive fallback to me after reviewing the changes of the third commit - please advise.

Further improvements for this would likely require either fundamental improvement of how detection polygons work to accurately model edges that are not simply straight lines between trace points, or a way to use the `SVGElement` to compute hover label positioning. I tried some approaches for the latter, but ultimately did not come up with a satisfying solution. I believe the current state is acceptable for the time being and a noticeable improvement over the current state.

_Reminder:_
_After opening a pull request, developer:_
- _should create a new small markdown log file using the PR number e.g. `1010_fix.md` or `1010_add.md` inside `draftlogs` folder as described in this [README](https://github.com/plotly/plotly.js/blob/master/draftlogs/README.md), commit it and push._
 - _should **not** force push (i.e. `git push -f`) to remote branches associated with opened pull requests. Force pushes make it hard for maintainers to keep track of updates. Therefore, if required, please fetch `upstream/master` and "merge" with master instead of "rebase"._

### CODE REVIEW

**Comment by lumip** [comment]
By the looks of it, this will likely also address #2399.

File: draftlogs/6865_fix.md [hunk 1]
```diff
@@ -0,0 +1 @@
+- Improved hover detection for for scatter plot fill tonext* [[#6865](https://github.com/plotly/plotly.js/pull/6865)]
```
Comment (archmoj) [inline draftlogs/6865_fix.md:1]:
```suggestion
- Improved hover detection for for scatter plot fill tonext* [[#6865](https://github.com/plotly/plotly.js/pull/6865)], with thanks to @lumip for the contribution!
```

**Comment by archmoj** [comment]
Looking excellent to me. :medal_military: :medal_sports: :1st_place_medal: 
Over to @alexcjohnson

**Review by alexcjohnson — APPROVED** [review]
💃 Looks great to me! tested by loading the `line-shape-arrow` mock and calling: 
```js
Plotly.restyle(gd,{fill:'tonexty',hoveron:'points+fills'})
```
On this branch it looks wayyy better than on master. Every once in a while I see the hover label jump around despite me not changing which shape I'm on top of, but it's still always labeling the correct object AFAICT. Very nicely done @lumip!
